lib/THC/THCHalf.cu

```diff
@@ -0,0 +1,105 @@
+#include "THCHalf.h"
+
+static const long NUM_THREADS = 256;
+static const long MAX_BLOCKS = 65535;
+
+__global__ void _floatToHalf_kernel(half *out, float* in, long len)
+{
+    long i = threadIdx.x + blockIdx.x * blockDim.x;
+    for (; i < len; i += gridDim.x * blockDim.x)
+        out[i] = __float2half(in[i]);
+}
+
+__global__ void _halfToFloat_kernel(float *out, half* in, long len)
+{
+    long i = threadIdx.x + blockIdx.x * blockDim.x;
+    for (; i < len; i += gridDim.x * blockDim.x)
+        out[i] = __half2float(in[i]);
+}
+
+void THCFloat2Half(THCState *state, half *out, float *in, long len) {
+    long num_blocks = (len + NUM_THREADS - 1) / NUM_THREADS;
+    if (num_blocks > MAX_BLOCKS)
+        num_blocks = MAX_BLOCKS;
+    _floatToHalf_kernel<<<num_blocks, NUM_THREADS, 0, THCState_getCurrentStream(state)>>>(out, in, len);
```

Comment:
use a thrust:: primitive here, rather than this kernel. they work out much better.
Look at thrust::transform http://docs.nvidia.com/cuda/thrust/#transformations